<h1> UK ROAD ACCIDENT DATA ANALYSIS<h1>
<h1>INCLUSIVE YEAR 2019 - 2022</h1>
<h1>Analyst: Jhon Lester Andal</h1>
<h3>DATA PREPARATION</h3>
<h4>Transfer Data to set Folder</h4>
<h3>Importinmg Libraries</h3>

In [1]:
import pandas as pd 
import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

<h1>DATA FRAME</h1>

In [2]:
accident = pd.read_csv('datasets//accident_data.csv')

In [3]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [4]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [5]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean())

In [6]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
accident[''] = accident['Longitude'].fillna(accident['Longitude'].mean())

In [8]:
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())

In [9]:
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

In [10]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])

In [11]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
                           0
dtype: int64

In [12]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
 14      

In [13]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident.dtypes
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

In [14]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
                           0
dtype: int64

In [15]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [16]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
                                  float64
dtype: object

In [17]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
                           0
dtype: int64

In [18]:
#accident[''] = accident[''].fillna(accident[''].mean())

In [19]:
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [20]:
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
                                  float64
dtype: object

In [21]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  category      
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

In [22]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
                           0
dtype: int64

In [23]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [24]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type            Year  Month  Day  \
0                                        Car -0.209082  2019      6    5   
1                                        Car -0.173647  2019      7    2   
2                      Taxi/Private hire car -0.210215  2019      8   26   
3       Bus or coach (17 or more pass seats) -0.202731  2019      8   16   
4                              Other vehicle -0.192487  2019      9    3   
...                                      ...       ...   ...    

In [25]:
 acci_19 = accident[accident['Year'] == 2019]

In [26]:
acci_19

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
182110       2.01E+12            Slight    2019-12-20  54.985289   
182111       2.01E+12           Serious    2019-12-21  54.984105   
182112       2.01E+12            Slight    2019-12-23  55.166369   
182113       2.01E+12            Slight    2019-12-23  54.995154   
182114       2.01E+12           Serious    2019-12-22  55.104618   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
182110       Darkness - no lighting   Dumfries and Galloway  -3.210294   
182111                     Daylight   Dumfries and Galloway  -3.193693   
182112       Darkness - no lighting   Dumfries and Galloway  -2.992068   
182113        Darkness - lights lit   Dumfries and Galloway  -3.058338   
182114                     Daylight   Dumfries and Galloway  -3.348857   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
182110                     1                   1            Frost or ice   
182111                     2                   1            Frost or ice   
182112                     1                   1            Frost or ice   
182113                     1                   1             Wet or damp   
182114                     2                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
182110  Single carriageway               Rural                  Other   
182111  Single carriageway               Rural                  Other   
182112  Single carriageway               Rural            Fog or mist   
182113  Single carriageway               Rural     Fine no high winds   
182114  Single carriageway               Rural  Raining no high winds   

                                Vehicle_Type            Year  Month  Day  \
0                                        Car -0.209082  2019      6    5   
1                                        Car -0.173647  2019      7    2   
2                      Taxi/Private hire car -0.210215  2019      8   26   
3       Bus or coach (17 or more pass seats) -0.202731  2019      8   16   
4                              Other vehicle -0.192487  2019      9    3   
...                                      ...       ...   ...    

In [27]:
acci_19['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

In [28]:
acci_19['Light_Conditions'].value_counts().sum()

182115

<h1>Insight 1</h1>
<h3> The accident mostly happen in dayligth which is 73.05%.</h3>

In [29]:
acci_19['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       148010
Raining no high winds     22835
Other                      4083
Raining + high winds       2959
Fine + high winds          2545
Fog or mist                 857
Snowing no high winds       680
Snowing + high winds        146
Name: count, dtype: int64

In [30]:
acci_19['Weather_Conditions'].value_counts().sum()

182115

<h1>Insight 2</h1>
<h3>In the record in fine no high winds weather codition have the most accident happen with 81%</h3>

In [31]:
casualty_19 = acci_19.groupby(['Weather_Conditions','Vehicle_Type'])['Number_of_Casualties'].count()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\1707058125.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  casualty_19 = acci_19.groupby(['Weather_Conditions','Vehicle_Type'])['Number_of_Casualties'].count()


In [32]:
casualty_19

Weather_Conditions     Vehicle_Type                        
Fine + high winds      Agricultural vehicle                       6
                       Bus or coach (17 or more pass seats)     157
                       Car                                     1817
                       Data missing or out of range               0
                       Goods 7.5 tonnes mgw and over             96
                                                               ... 
Snowing no high winds  Other vehicle                              2
                       Pedal cycle                                0
                       Ridden horse                               0
                       Taxi/Private hire car                     20
                       Van / Goods 3.5 tonnes mgw or under       24
Name: Number_of_Casualties, Length: 128, dtype: int64

In [33]:
df_exam3 = acci_19[acci_19['Weather_Conditions'] == 'Fine no high winds']

In [34]:
df_exam4 = df_exam3.groupby(['Vehicle_Type', 'Urban_or_Rural_Area'])['Number_of_Casualties'].size()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\3172274190.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam4 = df_exam3.groupby(['Vehicle_Type', 'Urban_or_Rural_Area'])['Number_of_Casualties'].size()


In [35]:
df_exam4.unstack()

Urban_or_Rural_Area                    Rural  Unallocated  Urban
Vehicle_Type                                                    
Agricultural vehicle                     177            0    323
Bus or coach (17 or more pass seats)    2870            1   5682
Car                                    38468            1  67412
Data missing or out of range               0            0      0
Goods 7.5 tonnes mgw and over           1819            0   3230
Goods over 3.5t. and under 7.5t          631            0   1062
Minibus (8 - 16 passenger seats)         168            0    334
Motorcycle 125cc and under              1112            0   2421
Motorcycle 50cc and under                603            0   1121
Motorcycle over 125cc and up to 500cc    671            0   1308
Motorcycle over 500cc                   2148            0   4075
Other vehicle                            491            0    828
Pedal cycle                               21            0     36
Ridden horse                               0            0      0
Taxi/Private hire car                   1146            0   2306
Van / Goods 3.5 tonnes mgw or under     2655            0   4890

In [36]:
df_exam3['Number_of_Casualties'].sum()
#total Number_of_Casualties in the weather 'fine with no winds' in 2019

199634

In [37]:

df_exam3.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()


C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\3245866668.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam3.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()


Urban_or_Rural_Area
Rural           78338
Unallocated         2
Urban          121294
Name: Number_of_Casualties, dtype: int64

<h1>Insight 3</h1>
<h3>total Number_of_Casualties in the weather 'fine with no winds' in urban areas in 2019 in 121294 or 60.75%</h3>

<h1>Insight 4</h1>
<h3>total Number_of_Casualties in the weather 'fine with no winds' in urban areas in 2019 in 78338 or 39.24%</h3>

<h1>Insight 5</h1>
<h3>the vehicle involve the most in accidents in the weather 'fine with no winds' in urban areas in 2019 is Car with 70% or 67412 involvement </h3>

<h1>Insight 6</h1>
<h3>the vehicle involve the most in accidents in the weather 'fine with no winds' in Rural areas in 2019 is  alsp Car  with 72.6% or 38468 involvement </h3>

In [38]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type            Year  Month  Day  \
0                                        Car -0.209082  2019      6    5   
1                                        Car -0.173647  2019      7    2   
2                      Taxi/Private hire car -0.210215  2019      8   26   
3       Bus or coach (17 or more pass seats) -0.202731  2019      8   16   
4                              Other vehicle -0.192487  2019      9    3   
...                                      ...       ...   ...    

In [39]:
accident['Accident_Severity'].unique()

['Serious', 'Slight', 'Fatal']
Categories (3, object): ['Fatal', 'Serious', 'Slight']

In [40]:
df_exam5 = accident[accident['Accident_Severity'] == 'Fatal']

In [41]:
df_exam5

Index Accident_Severity Accident Date   Latitude  \
7       200701BS67370             Fatal    2019-10-03  51.482260   
11      200701BS67747             Fatal    2019-11-02  51.499127   
20      200701BS68301             Fatal    2019-10-24  51.505996   
21      200701BS68302             Fatal    2019-11-24  51.486795   
717     200701CP00032             Fatal    2019-01-30  51.510978   
...               ...               ...           ...        ...   
660470  201091NH00529             Fatal    2022-01-29  58.475726   
660519  201091NJ01533             Fatal    2022-02-02  58.495554   
660565  201091NJ11329             Fatal    2022-08-15  57.871514   
660581  201091NJ14695             Fatal    2022-10-21  58.445475   
660627  201091NL00060             Fatal    2022-01-06  60.241195   

             Light_Conditions           District Area  Longitude  \
7       Darkness - lights lit  Kensington and Chelsea  -0.179486   
11                   Daylight  Kensington and Chelsea  -0.159652   
20                   Daylight  Kensington and Chelsea  -0.214277   
21                   Daylight  Kensington and Chelsea  -0.170375   
717                  Daylight          City of London  -0.107581   
...                       ...                     ...        ...   
660470               Daylight           Western Isles  -6.278583   
660519               Daylight                Highland  -4.451628   
660565               Daylight                Highland  -4.243081   
660581  Darkness - lights lit                Highland  -3.065535   
660627               Daylight        Shetland Islands  -1.409761   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
7                          3                   2                     Dry   
11                         1                   2                     Dry   
20                         1                   1                     Dry   
21                         5                   1                     Dry   
717                        1                   2             Wet or damp   
...                      ...                 ...                     ...   
660470                     1                   1             Wet or damp   
660519                     1                   1                    Snow   
660565                     1                   1                     Dry   
660581                     1                   1             Wet or damp   
660627                     3                   1                    Snow   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
7       Single carriageway               Urban     Fine no high winds   
11      Single carriageway               Urban     Fine no high winds   
20      Single carriageway               Urban     Fine no high winds   
21      Single carriageway               Urban     Fine no high winds   
717       Dual carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660470  Single carriageway               Rural      Fine + high winds   
660519  Single carriageway               Rural  Snowing no high winds   
660565  Single carriageway               Rural     Fine no high winds   
660581  Single carriageway               Rural     Fine no high winds   
660627  Single carriageway               Rural     Fine no high winds   

                                Vehicle_Type            Year  Month  Day  \
7                                        Car -0.179486  2019     10    3   
11      Bus or coach (17 or more pass seats) -0.159652  2019     11    2   
20                                       Car -0.214277  2019     10   24   
21                                       Car -0.170375  2019     11   24   
717                                      Car -0.107581  2019      1   30   
...                                      ...       ...   ...    ...  ...   
660470                                   Car -6.278583  2022

In [42]:
df_exam6 = df_exam5.groupby(['Light_Conditions'])['Number_of_Casualties'].sum()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\4028298760.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam6 = df_exam5.groupby(['Light_Conditions'])['Number_of_Casualties'].sum()


In [43]:
df_exam5['Number_of_Casualties'].sum()
# total number of Accident_Severity = fatal

16483

In [44]:
df_exam6
#Accident_Severity = fatal

Light_Conditions
Darkness - lighting unknown     114
Darkness - lights lit          3421
Darkness - lights unlit          76
Darkness - no lighting         3362
Daylight                       9510
Name: Number_of_Casualties, dtype: int64

In [45]:
df_exam5.groupby(['Light_Conditions', 'Road_Surface_Conditions']).size()


C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\2498991274.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam5.groupby(['Light_Conditions', 'Road_Surface_Conditions']).size()


Light_Conditions             Road_Surface_Conditions
Darkness - lighting unknown  Dry                          43
                             Flood over 3cm. deep          1
                             Frost or ice                  1
                             Snow                          0
                             Wet or damp                  23
Darkness - lights lit        Dry                        1076
                             Flood over 3cm. deep          5
                             Frost or ice                 43
                             Snow                          5
                             Wet or damp                 731
Darkness - lights unlit      Dry                          26
                             Flood over 3cm. deep          1
                             Frost or ice                  0
                             Snow                          0
                             Wet or damp                  18
Darkness - no lighting       Dry

<h1>Insight 7</h1>
<h3> Accident_Severity = fatal also more happening in day time with 57.69% or 9510 out of 16483</h3>

In [46]:
df_exam7 = df_exam5.groupby(['Light_Conditions','Road_Surface_Conditions'])['Number_of_Casualties'].sum()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\2929783886.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam7 = df_exam5.groupby(['Light_Conditions','Road_Surface_Conditions'])['Number_of_Casualties'].sum()


In [47]:
df_exam7.unstack()

Road_Surface_Conditions       Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Light_Conditions                                                              
Darkness - lighting unknown    71                     3             3     0   
Darkness - lights lit        1940                    13            97    10   
Darkness - lights unlit        43                     5             0     0   
Darkness - no lighting       1696                    10           173    19   
Daylight                     6823                    21           174    42   

Road_Surface_Conditions      Wet or damp  
Light_Conditions                          
Darkness - lighting unknown           37  
Darkness - lights lit               1361  
Darkness - lights unlit               28  
Darkness - no lighting              1464  
Daylight                            2450

In [48]:
df_exam7['Daylight'].sum()

9510

<h1>Insight 8</h1>
<h3> Accident_Severity that are fatal more happening in day time and dry condition with 43.81%</h3>

<h1>Insight 9</h1>
<h3> Accident_Severity that are  fatal are less happening when theres a Flood over 3cm. deep with only 0.26%</h3>

In [49]:
acci_20 = accident[accident['Year'] == 2020]

In [50]:
acci_20['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     113716
Wet or damp              52076
Frost or ice              3870
Snow                       598
Flood over 3cm. deep       331
Name: count, dtype: int64

In [51]:
acci_20['Number_of_Casualties'].sum()

230905

<h1>Insight 10</h1>
<h3> The number of accident more happening in dry road type condition with 49.24% in 2020 </h3>

In [52]:
df_exam8 = acci_20.groupby(['Road_Surface_Conditions','Number_of_Vehicles'])['Number_of_Casualties'].sum()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\683350287.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_exam8 = acci_20.groupby(['Road_Surface_Conditions','Number_of_Vehicles'])['Number_of_Casualties'].sum()


In [53]:
acci_20['Number_of_Vehicles'].sum()

311604

In [54]:
df_exam8.unstack()

Number_of_Vehicles          1      2      3     4    5    6    7   8   9   10  \
Road_Surface_Conditions                                                         
Dry                      37975  93190  14764  3794  916  307  125  75  11   2   
Flood over 3cm. deep       225    231     41     3    4    0    0   0   0   0   
Frost or ice              2437   2320    340    80   25    9    2  11   1   5   
Snow                       344    358     68    34    3    3    9   0   0   6   
Wet or damp              20234  42487   7718  1863  459  201   94  40  20  18   

Number_of_Vehicles       11  12  14  16  
Road_Surface_Conditions                  
Dry                       0   0  11   0  
Flood over 3cm. deep      0   0   0   0  
Frost or ice              0   0   0  17  
Snow                      0   0   3   0  
Wet or damp              16   6   0   0

In [55]:
df_exam8['Dry'].sum()

151170

<h1>Insight 11</h1>
<h3> The number of casualty when the Number_of_Vehicles is 2 with 29.90% in dry road surface in 2020 </h3>

In [56]:
df_exam9 = acci_20[acci_20['Road_Surface_Conditions'] == 'Frost or ice']

In [57]:
df_exam9['Number_of_Casualties'].sum()


5247

<h1>Insight 12</h1>
<h3> The number of casualty in road surface condition frost or ice  in 2020 is 5247 that 2.27% of 230905(total casualy in 2020)</h3>

<h1>Insight 13</h1>
<h3>The number of casualty in Number_of_Vehicles 1 vehicle have more than 2 vehicle in road surface frost or ice. In 1 vehicle 46.44% and in 2 vehicle with 44.21%</h3>

In [58]:
df_Exam10 =  acci_20.groupby(['Road_Surface_Conditions','Vehicle_Type'])['Number_of_Casualties'].sum()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\1704661901.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_Exam10 =  acci_20.groupby(['Road_Surface_Conditions','Vehicle_Type'])['Number_of_Casualties'].sum()


In [59]:
df_Exam10.unstack()

Vehicle_Type             Agricultural vehicle  \
Road_Surface_Conditions                         
Dry                                       477   
Flood over 3cm. deep                        7   
Frost or ice                               20   
Snow                                        2   
Wet or damp                               213   

Vehicle_Type             Bus or coach (17 or more pass seats)     Car  \
Road_Surface_Conditions                                                 
Dry                                                      5972  113428   
Flood over 3cm. deep                                       21     366   
Frost or ice                                              232    3937   
Snow                                                       39     615   
Wet or damp                                              2952   54579   

Vehicle_Type             Data missing or out of range  \
Road_Surface_Conditions                                 
Dry                                                 3   
Flood over 3cm. deep                                0   
Frost or ice                                        2   
Snow                                                0   
Wet or damp                                         4   

Vehicle_Type             Goods 7.5 tonnes mgw and over  \
Road_Surface_Conditions                                  
Dry                                               3982   
Flood over 3cm. deep                                12   
Frost or ice                                       112   
Snow                                                31   
Wet or damp                                       1939   

Vehicle_Type             Goods over 3.5t. and under 7.5t  \
Road_Surface_Conditions                                    
Dry                                                 1377   
Flood over 3cm. deep                                   0   
Frost or ice                                          53   
Snow                                                   5   
Wet or damp                                          683   

Vehicle_Type             Minibus (8 - 16 passenger seats)  \
Road_Surface_Conditions                                     
Dry                                                   497   
Flood over 3cm. deep                                    0   
Frost or ice                                           10   
Snow                                                    5   
Wet or damp                                           219   

Vehicle_Type             Motorcycle 125cc and under  \
Road_Surface_Conditions                               
Dry                                            3549   
Flood over 3cm. deep                              8   
Frost or ice                                    129   
Snow                                             20   
Wet or damp                                    1763   

Vehicle_Type             Motorcycle 50cc and under  \
Road_Surface_Conditions                              
Dry                                           1558   
Flood over 3cm. deep                             6   
Frost or ice                                    53   
Snow                                            12   
Wet or damp                                    755   

Vehicle_Type             Motorcycle over 125cc and up to 500cc  \
Road_Surface_Conditions                                          
Dry                                                       1676   
Flood over 3cm. deep                                         7   
Frost or ice                                                67   
Snow                                                         9   
Wet or damp                                                795   

Vehicle_Type             Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Road_Surface_Conditions                                                      
Dry                                       6006           1401           63   
Flood over 3cm. deep      

<h1>Insight 14</h1>
<h3>Theres no casualty in riding horse in any road surface condition</h3>

In [60]:
acci_20

Index Accident_Severity Accident Date   Latitude  \
182115  200801BS69439           Serious    2020-01-23  51.506812   
182116  200801BS69594           Serious    2020-02-15  51.496323   
182117  200801BS69698           Serious    2020-02-27  51.502042   
182118  200801BS69935           Serious    2020-02-25  51.492733   
182119  200801BS69938           Serious    2020-02-27  51.493271   
...               ...               ...           ...        ...   
352701       2.01E+12            Slight    2020-11-24  55.233969   
352702       2.01E+12            Slight    2020-12-06  54.983911   
352703       2.01E+12            Slight    2020-12-17  55.008072   
352704       2.01E+12            Slight    2020-12-16  54.986388   
352705       2.01E+12            Slight    2020-12-19  55.067150   

              Light_Conditions           District Area  Longitude  \
182115   Darkness - lights lit  Kensington and Chelsea  -0.214677   
182116                Daylight  Kensington and Chelsea  -0.170138   
182117   Darkness - lights lit  Kensington and Chelsea  -0.190946   
182118                Daylight  Kensington and Chelsea  -0.193763   
182119   Darkness - lights lit  Kensington and Chelsea  -0.199504   
...                        ...                     ...        ...   
352701  Darkness - no lighting   Dumfries and Galloway  -3.394276   
352702                Daylight   Dumfries and Galloway  -3.195094   
352703  Darkness - no lighting   Dumfries and Galloway  -3.334825   
352704                Daylight   Dumfries and Galloway  -3.180789   
352705                Daylight   Dumfries and Galloway  -3.403817   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
182115                     1                   1                     Dry   
182116                     1                   1                     Dry   
182117                     1                   1                     Dry   
182118                     1                   1                     Dry   
182119                     1                   2                     Dry   
...                      ...                 ...                     ...   
352701                     1                   1            Frost or ice   
352702                     1                   1            Frost or ice   
352703                     1                   1             Wet or damp   
352704                     1                   1                     Dry   
352705                     1                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area  Weather_Conditions  \
182115  Single carriageway               Urban  Fine no high winds   
182116    Dual carriageway               Urban  Fine no high winds   
182117  Single carriageway               Urban  Fine no high winds   
182118          Roundabout               Urban  Fine no high winds   
182119    Dual carriageway               Urban  Fine no high winds   
...                    ...                 ...                 ...   
352701  Single carriageway               Rural  Fine no high winds   
352702  Single carriageway               Rural  Fine no high winds   
352703  Single carriageway               Rural  Fine no high winds   
352704  Single carriageway               Rural  Fine no high winds   
352705  Single carriageway               Rural  Fine no high winds   

                                Vehicle_Type            Year  Month  Day  \
182115                                   Car -0.214677  2020      1   23   
182116  Bus or coach (17 or more pass seats) -0.170138  2020      2   15   
182117   Van / Goods 3.5 tonnes mgw or under -0.190946  2020      2   27   
182118                                   Car -0.193763  2020      2   25   
182119                 Motorcycle over 500cc -0.199504  2020      2   27   
...                                      ...       ...   ...    ...  ...   
352701      Minibus (8 - 16 passenger seats) -3.394276  2020     11   24   
352702  

In [61]:
df_Exam11 =  acci_20.groupby(['Accident_Severity','Vehicle_Type'])['Number_of_Casualties'].sum()

C:\Users\RIZEN\AppData\Local\Temp\ipykernel_5004\474699759.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_Exam11 =  acci_20.groupby(['Accident_Severity','Vehicle_Type'])['Number_of_Casualties'].sum()


In [62]:
df_Exam11.unstack()

Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                                 8                                   250   
Serious                              92                                  1245   
Slight                              619                                  7721   

Vehicle_Type          Car  Data missing or out of range  \
Accident_Severity                                         
Fatal                3263                             0   
Serious             25621                             0   
Slight             144041                             9   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                        113   
Serious                                      906   
Slight                                      5057   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                           38   
Serious                                        339   
Slight                                        1741   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                            18   
Serious                                         102   
Slight                                          611   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                      74                         38   
Serious                                   788                        347   
Slight                                   4607                       1999   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                                 38   
Serious                                              353   
Slight                                              2163   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                                170             38            3   
Serious                             1321            302           22   
Slight                              7732           1770           62   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                         0                    111   
Serious                       0                    746   
Slight                        0                   4090   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                              236  
Serious                                           1803  
Slight                                           10298

In [63]:
accident['Day'].unique()

array([ 5,  2, 26, 16,  3, 18, 31,  9, 12,  8,  1, 19, 10, 20, 24, 22, 17,
        6, 13,  7, 15, 14,  4, 28, 27, 25, 30, 11, 21, 23, 29])

<h1>Insight 15</h1>
<h3>In all accident_severity(fatal, serious, slight) the vehicle car have the most number of people involve</h3>

In [64]:
data = acci_20.groupby(['Month'])['Number_of_Casualties'].sum()


In [65]:
data

Month
1     18940
2     19189
3     18337
4     18286
5     19422
6     19129
7     19761
8     18680
9     19388
10    21029
11    19905
12    18839
Name: Number_of_Casualties, dtype: int64

In [66]:
acci_20.groupby('Month')['Number_of_Casualties'].sum().sum()


230905

<h1>Insight 16</h1>
<h3>October has the highest number of casualties in 2020 with 9.10%</h3>

In [67]:
data = acci_20.groupby(['Month'])['Number_of_Casualties'].size()


In [68]:
data


Month
1     14133
2     14353
3     13494
4     13394
5     14336
6     14205
7     14630
8     13366
9     14432
10    15684
11    14770
12    13794
Name: Number_of_Casualties, dtype: int64

In [69]:
data.min()

13366

In [70]:
data.max()

15684

In [71]:
acci_20['Number_of_Casualties'].value_counts().sum()

170591

<h1>Insight 17</h1>
<h3>Also in October have the highest accident rate  in 2020 with 9.19%</h3>

<h1>Insight 18</h1>
<h3>The accident rate are ranging from 13300 - 15700 in year 2020</h3>

In [72]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type            Year  Month  Day  \
0                                        Car -0.209082  2019      6    5   
1                                        Car -0.173647  2019      7    2   
2                      Taxi/Private hire car -0.210215  2019      8   26   
3       Bus or coach (17 or more pass seats) -0.202731  2019      8   16   
4                              Other vehicle -0.192487  2019      9    3   
...                                      ...       ...   ...    

In [73]:
correlation1 = accident['Number_of_Vehicles'].corr(accident['Number_of_Casualties'])

print(correlation1)

0.22888886126927557


<h1>Insight 19</h1>
<h3>Theres no correlation between the Number_of_Vehicles and Number_of_Casualties</h3>

In [80]:
accident['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
2     391995
1     200787
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
14         9
12         7
13         6
16         4
28         1
15         1
32         1
19         1
Name: count, dtype: int64

In [81]:
accident['Number_of_Vehicles'].value_counts().sum()

660679

<h1>Insight 20</h1>
<h3>Most accident involve 2 vehicle with 59.24%</h3>